evm/src/cpu/kernel/stack_manipulation.rs

```diff
@@ -0,0 +1,259 @@
+
+    // For each node, stores `(best_cost, Option<(parent, op)>)`.
+    let mut node_info = HashMap::<Vec<StackItem>, (u32, Option<(Vec<StackItem>, StackOp)>)>::new();
+    node_info.insert(src.clone(), (0, None));
+
+    while let Some(node) = queue.pop() {
+        if node.stack == dst {
+            // The destination is now the lowest-cost node, so we must have found the best path.
+            let mut path = vec![];
+            let mut stack = &node.stack;
+            // Rewind back to src, recording a list of operations which will be backwards.
+            while let Some((parent, op)) = &node_info[stack].1 {
+                stack = parent;
+                path.push(op.clone());
+            }
+            assert_eq!(stack, &src);
+            path.reverse();
+            return path;
+        }
+
+        let (best_cost, _) = node_info[&node.stack];
+        if best_cost < node.cost {
+            // Since we can't efficiently remove nodes from the heap, it can contain duplicates.
+            // In this case, we've already visited this stack state with a lower cost.
+            continue;
+        }
+
+        for op in next_ops(&node.stack, &dst, &unique_literals) {
+            let neighbor = match op.apply_to(node.stack.clone()) {
+                Some(n) => n,
+                None => continue,
+            };
+
+            let cost = node.cost + op.cost();
+            let entry = node_info.entry(neighbor.clone());
+            if let Occupied(e) = &entry && e.get().0 <= cost {
+                // We already found a better or equal path.
+                continue;
+            }
+
+            let neighbor_info = (cost, Some((node.stack.clone(), op.clone())));
+            match entry {
+                Occupied(mut e) => {
+                    e.insert(neighbor_info);
+                }
+                Vacant(e) => {
+                    e.insert(neighbor_info);
+                }
+            }
+
+            queue.push(Node {
+                stack: neighbor,
+                cost,
+            });
+        }
+    }
+
+    panic!("No path found from {:?} to {:?}", src, dst)
+}
+
+/// A node in the priority queue used by Dijkstra's algorithm.
+#[derive(Eq, PartialEq)]
+struct Node {
+    stack: Vec<StackItem>,
+    cost: u32,
+}
+
+impl PartialOrd for Node {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Node {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // We want a min-heap rather than the default max-heap, so this is the opposite of the
+        // natural ordering of costs.
+        other.cost.cmp(&self.cost)
+    }
+}
+
+/// Like `StackReplacement`, but without constants or macro vars, since those were expanded already.
+#[derive(Eq, PartialEq, Hash, Clone, Debug)]
+enum StackItem {
+    NamedItem(String),
+    Literal(Literal),
+}
+
+#[derive(Clone, Debug)]
+enum StackOp {
+    Push(Literal),
+    Pop,
+    Dup(u8),
+    Swap(u8),
+}
```

Comment:
Not sure if we'll need it, but could be also cool in the future to support MLOAD either with a constant offset or one present on the stack.

Reply:
Ah like using memory to help rearrange the stack? Could be interesting though using `MLOAD_GENERAL` may be a bit expensive since we need to push the context/segment